### PR TITLE
device: use hash of UID registers to generate unique product number

### PR DESCRIPTION
**Pull request by mike-scott**
For each device there are a set of unique device ID registers.  These
vary in length.  Let's generate a hash of the combined hex strings
of these registers to use as a unique product number.

Tested on: K64F, Carbon and Nitrogen

Signed-off-by: Michael Scott <michael.scott@linaro.org>

### CODE REVIEW

**Comment by gatek33p3r** [comment]
Checkpatch Failed.

**Comment by gatek33p3r** [comment]
Bootloader Build Succeeded.

**Comment by gatek33p3r** [comment]
Application Build Succeeded.

**Comment by gatek33p3r** [comment]
Checkpatch Suceeded.

**Comment by gatek33p3r** [comment]
Bootloader Build Succeeded.

**Comment by gatek33p3r** [comment]
Application Build Succeeded.

**Comment by gatek33p3r** [comment]
Checkpatch Suceeded.

**Comment by gatek33p3r** [comment]
Application Build Succeeded.

**Comment by gatek33p3r** [comment]
Bootloader Build Succeeded.

**Comment by gatek33p3r** [comment]
Checkpatch Suceeded.

**Comment by gatek33p3r** [comment]
Application Build Succeeded.

**Comment by gatek33p3r** [comment]
Bootloader Build Succeeded.

File: src/device.h [hunk 1]
```diff
@@ -48,8 +48,15 @@
 #define UNKNOWN_APPEARANCE 0x0000
 
 /* General hardware specific configs */
-#if defined(CONFIG_SOC_SERIES_STM32F4X)
-#define DEVICE_ID_BASE	0x1fff7a10
+#if defined(CONFIG_SOC_SERIES_NRF52X)
+#define DEVICE_ID_BASE		(NRF_FICR_BASE + 0x60)
```
Comment (mbolivar) [inline src/device.h]: Maybe something like: `&NRF_FICR->DEVICEID[0]`?

File: src/device.h [hunk 2]
```diff
@@ -48,8 +48,15 @@
...
+#define DEVICE_ID_LENGTH	2
```
Comment (mbolivar) [inline src/device.h]: `ARRAY_SIZE(NRF_FICR->DEVICEID) / sizeof(uint32_t)`, maybe? Or just a comment saying these are measured in 32 bit words.

File: src/device.h (hunk 1 above)
Comment (mike-scott) [inline src/device.h]: ACK.  Good idea.  Will change.

File: src/device.h (hunk 2 above)
Comment (mike-scott) [inline src/device.h]: ACK.  Adding a comment that these are currently 32-bit words

**Comment by gatek33p3r** [comment]
Checkpatch Failed.

**Comment by gatek33p3r** [comment]
Bootloader Build Succeeded.

**Comment by gatek33p3r** [comment]
Application Build Succeeded.

**Comment by gatek33p3r** [comment]
Checkpatch Suceeded.

**Comment by gatek33p3r** [comment]
Bootloader Build Succeeded.

**Comment by gatek33p3r** [comment]
Application Build Succeeded.

File: src/device.c [hunk 3]
```diff
@@ -13,17 +14,31 @@ struct product_id_t product_id = {
...
-				((*(uint8_t *) DEVICE_ID_BASE + 2) << 8) +
-				((*(uint8_t *) DEVICE_ID_BASE + 4) << 16) +
-				((*(uint8_t *) DEVICE_ID_BASE + 7) << 24);
-#endif
+	int i;
+	char buffer[DEVICE_ID_LENGTH*8 + 1];
+
+	for (i = 0; i < DEVICE_ID_LENGTH; i++) {
+		sprintf(buffer + i*8, "%08x",
+			*(((uint32_t *)DEVICE_ID_BASE) + i));
```
Comment (mike-scott) [inline src/device.c]: NOTE: I typecast DEVICE_ID_BASE prior to doing any math on it to make sure that we're iterating in 32-bit addresses (previously the code assumed a long int pointer to bytes and used i*4 which was incorrect in some cases depending on the definition of DEVICE_ID_BASE).

**Comment by gatek33p3r** [comment]
Checkpatch Suceeded.

**Comment by gatek33p3r** [comment]
Application Build Failed.

**Comment by gatek33p3r** [comment]
Bootloader Build Failed.

**Comment by gatek33p3r** [comment]
Checkpatch Suceeded.

**Comment by gatek33p3r** [comment]
Application Build Succeeded.

**Comment by gatek33p3r** [comment]
Bootloader Build Succeeded.

**Pull request merged**